### PR TITLE
Add Auto-Scaling and other Features

### PR DESCRIPTION
**Pull request by SubtleDragon057**
This update includes auto-scaling, playmat UI and functions, remove card effects, fix to Dark Magician model animation, new layers to assist with the added effects, and a button to return to connection screen.

### CODE REVIEW

File: Assets/Code/Core/General/AnimatorIDSetter.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+using UnityEngine;
+
+namespace AssemblyCSharp.Assets.Code.Core.General
+{
+    public static class AnimatorIDSetter
+    {
+        #region Monster Animations
+        public static readonly int Animator_Summoning_Trigger = Animator.StringToHash("SummoningTrigger");
```
Comment (BramDC3) [inline Assets/Code/Core/General/AnimatorIDSetter.cs:8]: There should be a blank line below #region and before #endregion

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+namespace AssemblyCSharp.Assets.Code.Core.General.Extensions
+{
+    public static class UtilityExtensions
+    {
+        public static float Map(this float value,
```
Comment (BramDC3) [inline Assets/Code/Core/General/Extensions/UtilityExtensions.cs:5]: The name of this function does not clearly imply what it does

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        public static float Map(this float value, 
+                                float originalMin, 
+                                float originalMax, 
+                                float newMin, 
+                                float newMax)
+        {
+            return (value - originalMin) / (originalMax - originalMin) * (newMax - newMin) + newMin;
+        }
+
+        public static bool IsWithin(this float value, float min, float max)
```
Comment (BramDC3) [inline Assets/Code/Core/General/Extensions/UtilityExtensions.cs:14]: Would rename this to "IsWithinRange"

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            return value >= min && value <= max;
+        }
+        
+        public static bool CheckIfTrue(string data)
```
Comment (BramDC3) [inline Assets/Code/Core/General/Extensions/UtilityExtensions.cs:19]: This function doesn't belong here and it's name doesn't imply what the function does

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            if(data == "0")
+            {
+                return true;
+            }
+            else if (data == "1")
```
Comment (BramDC3) [inline Assets/Code/Core/General/Extensions/UtilityExtensions.cs:25]: No need for else if and else as you're returning values

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+            {
+                return false;
+            }
+            else { return false; }
```
Comment (BramDC3) [inline Assets/Code/Core/General/Extensions/UtilityExtensions.cs:29]: Bad formatting

File: Assets/Animations/Loading - Eye of Ra.anim [hunk 7]
```diff
@@ -0,0 +1,307 @@
+%YAML 1.1
```
Comment (BramDC3) [inline Assets/Animations/Loading - Eye of Ra.anim:1]: Can be removed I guess?

File: Assets/Code/Features/SpeedDuel/SpeedDuelView.cs [hunk 8]
```diff
@@ -130,6 +141,58 @@ private void PlaceObject()
             SpeedDuelField = Instantiate(_objectToPlace, _placementPose.position, _placementPose.rotation);
         }
 
+        private void SetPendulumScale(List<ARRaycastHit> hits)
```
Comment (BramDC3) [inline Assets/Code/Features/SpeedDuel/SpeedDuelView.cs:144]: Wait, what? :p

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs (hunk 2 above)
Comment (SubtleDragon057) [reply Assets/Code/Core/General/Extensions/UtilityExtensions.cs:5]: Renamed to "MapToNewValue". Map is part of the Arduino library, hopefully this name is clearer to understand.

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs (hunk 3 above)
Comment (SubtleDragon057) [reply Assets/Code/Core/General/Extensions/UtilityExtensions.cs:14]: Changed to suggested name

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs (hunk 4 above)
Comment (SubtleDragon057) [reply Assets/Code/Core/General/Extensions/UtilityExtensions.cs:19]: This wasn't supposed to be added with this pull request. Removed until it's being used

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs (hunk 5 above)
Comment (SubtleDragon057) [reply Assets/Code/Core/General/Extensions/UtilityExtensions.cs:25]: Common sense sometimes escapes me, fixed this for when this function is used

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs (hunk 6 above)
Comment (SubtleDragon057) [reply Assets/Code/Core/General/Extensions/UtilityExtensions.cs:29]: Whoops, habits from the Arduino IDE. Fixed formatting for when this function is used

File: Assets/Animations/Loading - Eye of Ra.anim (hunk 7 above)
Comment (SubtleDragon057) [reply Assets/Animations/Loading - Eye of Ra.anim:1]: Yes, I must've kept this accidentally when fixing the merge conflicts. Deleted.

File: Assets/Code/Features/SpeedDuel/SpeedDuelView.cs (hunk 8 above)
Comment (SubtleDragon057) [reply Assets/Code/Features/SpeedDuel/SpeedDuelView.cs:144]: I see no problem with this function name xD was originally changed from "SetPlaymatScale"

File: Assets/Code/Core/General/AnimatorIDSetter.cs (hunk 1 above)
Comment (SubtleDragon057) [reply Assets/Code/Core/General/AnimatorIDSetter.cs:9]: Fixed Formatting

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+namespace AssemblyCSharp.Assets.Code.Core.General.Extensions
+{
+    public static class UtilityExtensions
+    {
+        public static float MapToNewValue(this float value,
```
Comment (BramDC3) [inline Assets/Code/Core/General/Extensions/UtilityExtensions.cs:5]: I think I already mentioned it, but I still don't think the name of this function is clear

File: Assets/Code/Core/General/Navigation/ReturnToMainMenu.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+using UnityEngine.UI;
+using Zenject;
+using UniRx;
+
+namespace AssemblyCSharp.Assets.Code.Core.General.Navigation
+{
+    public class ReturnToMainMenu : MonoBehaviour
+    {
+        [SerializeField]
+        private Button _backButton;
```
Comment (BramDC3) [inline Assets/Code/Core/General/Navigation/ReturnToMainMenu.cs:12]: We should only have UI elements in views. I'm not sure for which screen this is, but it might be better to make a view for it and inject the navigation service in the view.

File: Assets/Code/Features/SpeedDuel/SpeedDuelView.cs (hunk 8 above)
Comment (BramDC3) [reply Assets/Code/Features/SpeedDuel/SpeedDuelView.cs:144]: Would rename it back to SetPlaymatScale :)

File: Assets/Code/Features/SpeedDuel/SpeedDuelView.cs [hunk 11]
```diff
@@ -102,12 +111,14 @@ private void UpdatePlacementPose()
             _placementPoseIsValid = hits.Count > 0;
             if (_placementPoseIsValid)
             {
-                _placementPose = hits[0].pose;
+                _placementPose = hits[hits.Count-1].pose;
```
Comment (BramDC3) [inline Assets/Code/Features/SpeedDuel/SpeedDuelView.cs:114]: Why was this changed?

File: Assets/Code/Core/General/Navigation/ReturnToMainMenu.cs (hunk 10 above)
Comment (SubtleDragon057) [reply Assets/Code/Core/General/Navigation/ReturnToMainMenu.cs:12]: I'm not sure what you mean by views, so I may need some clarification, however I can definitely move it. This button is reused on any screens past the main menu in order to return to the main menu (ie. speed duel and model viewer)

File: Assets/Code/Features/SpeedDuel/SpeedDuelView.cs (hunk 11 above)
Comment (SubtleDragon057) [reply Assets/Code/Features/SpeedDuel/SpeedDuelView.cs:114]: The first index of a ray cast hits array is the closest detected object. I found that it began tracking planes in mid-air when moving the mat around. Count - 1 gives the last ray cast hit, which returns the actual surface the phone is pointed at, allowing the mat to be properly placed on the ground.

File: Assets/Code/Features/SpeedDuel/SpeedDuelView.cs (hunk 8 above)
Comment (SubtleDragon057) [reply Assets/Code/Features/SpeedDuel/SpeedDuelView.cs:144]: I'll concede...but I don't like it xD

File: Assets/Code/Core/General/Extensions/UtilityExtensions.cs (hunk 9 above)
Comment (SubtleDragon057) [reply Assets/Code/Core/General/Extensions/UtilityExtensions.cs:5]: Would a name like "RemapToNewRangeOfValues" work better? This function is used to assist in smooth UI sliders when dealing with decimal numbers (ie. the playmat scale)

File: Assets/Code/Features/SpeedDuel/PlaymatViewLogic.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
+using UnityEngine;
+using UnityEngine.UI;
+using AssemblyCSharp.Assets.Code.Core.General;
+
+namespace AssemblyCSharp.Assets.Code.Features.SpeedDuel
+{
+    public class PlaymatViewLogic : MonoBehaviour
+    {
+        private static readonly string Indicator_Tag = "Indicator";
```
Comment (BramDC3) [inline Assets/Code/Features/SpeedDuel/PlaymatViewLogic.cs:9]: Should be named INDICATOR_TAG

File: Assets/Code/Features/SpeedDuel/PlaymatViewLogic.cs [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+            _animators = GetComponentsInChildren<Animator>();
+            //TODO Create a more solid connection to Indicator object
+            _interaction = GameObject.FindGameObjectWithTag(Indicator_Tag);
+        }
+
+        public void SetScaleStartPosition(float startPosition) => _scaleSlider.value = startPosition;
+
+        public void ScalePlaymat(float scale) 
+        {
+            var mappedScale = MapToNewSliderValueToNewValue(scale, _scaleSlider.minValue, _scaleSlider.maxValue, 0.1f, 5f);
```
Comment (BramDC3) [inline Assets/Code/Features/SpeedDuel/PlaymatViewLogic.cs:34]: Map ToNewSliderValue ToNewValue? :p

Comment (SubtleDragon057) [reply Assets/Code/Features/SpeedDuel/PlaymatViewLogic.cs:34]: Huh, guess I should proofread first xD
